### PR TITLE
Improve tap-snapshot folder structure

### PR DESCRIPTION
**Pull request by isaacs**
Previously, every test would save its snapshots in a file in
`tap-snapshots` with a name that was the sanitized `t.fullname`.  And
since each test uses the snapshot file from its parent test, the
test.name was almost always 'TAP'.

This led to a lot of files named things like
`tap-snapshots/test-lib-utils-base-utils.js-TAP.test.cjs`.  When
snapshots had to be examined or reasoned about, it made it unnecessarily
difficult.

This changes the naming convention to instead take the relative path
from the current working directory to the main script as-is, and root it
on the `tap-snapshots` folder.  Then, any arguments beyond the script
name are sanitized and appended, and finally the `.test.cjs` extension
is added.

Thus, for the example above, you get
`tap-snapshots/test/lib/utils/base.js.test.cjs` as the snapshot name for
`test/lib/utils/base.js`, which makes the folder _much_ more covenient
to work with on larger projects, and less repetitive in the filenames.

### CODE REVIEW

File: lib/test.js [hunk 1]
```diff
@@ -1174,8 +1174,15 @@ class Test extends Base {
...
+  get snapshotFile () {
+    return this[_getSnapshot]().file
+  }
+  set snapshotFile (file) {
+    this[_getSnapshot]().file = file
+  }
+
   get testdirName () {
-    const re = /[^a-z0-9]+/ig
+    const re = /[^a-zA-Z0-9\._\-]+/ig
```
Comment (coreyfarrell) [inline lib/test.js:1185]: At some point we probably want to separate a `./sanitize-filename.js` function (DRY).

File: lib/snapshot.js [hunk 2]
```diff
@@ -22,6 +29,10 @@ class Snapshot {
     const index = +this.indexes.get(message) || 1
     this.indexes.set(message, index + 1)
     try {
+      // throw before the require() for missing file, because node
+      // caches module load errors, and we might create it at some point.
```
Comment (coreyfarrell) [inline lib/snapshot.js:33]: Would it be better to `delete require.cache[this.file]` before the require?

Comment (isaacs) [reply lib/snapshot.js:33]: Eh.  I prefer to just work around known/documented Node.js semantics rather than try to get cute with circumventing them.

Comment (isaacs) [reply lib/snapshot.js:33]: (Wouldn't know that from many of my OSS modules, but I do try to learn from my past sins 😂)

**Pull request merged**